### PR TITLE
Allow RST_STREAM to be sent and received after FIN

### LINKED ISSUE #779: Is it permitted to send both FIN and RST
@mcmanus argues that this is useful because you might send FIN and then get tired of waiting for ACKs and that it's processed on the receiver side where first one that comes ready (RST received, FIN in-order) is processed and the other ignored. My reading of the text is that this is implicitly allowed.

### PR DESCRIPTION
**Pull request by janaiyengar**
Addresses #779.

### CODE REVIEW

File: draft-ietf-quic-transport.md [hunk 1]
```diff
@@ -2648,6 +2650,10 @@ closes a stream SHOULD be discarded.  An endpoint MAY choose to limit the period
 over which it ignores frames and treat frames that arrive after this time as
 being in error.
 
+An endpoint may receive a RST_STREAM after a FIN if the peer resets the stream
+after sending a FIN on it. In this case, the endpoint MAY discard any data that
+it already received on that stream.
```
Comment (ianswett) [inline draft-ietf-quic-transport.md]: How about adding a sentence that the endpoing MAY close the connection if a stream is reset with a different offset from the previously received fin.

File: draft-ietf-quic-transport.md [hunk 2]
```diff
@@ -2615,8 +2615,10 @@ Any frame type that mentions a stream ID can be sent in this state.
 
 A stream that is in the "half-closed (local)" state MUST NOT be used for sending
 on new STREAM frames.  Retransmission of data that has already been sent on
-STREAM frames is permitted.  An endpoint MAY also send MAX_STREAM_DATA and
-STOP_SENDING in this state.
+STREAM frames is permitted.  If the stream is canceled or reset while in this
+state, a RST_STREAM frame MAY be sent. A RST_STREAM frame sent in this state
```
Comment (ekr) [inline draft-ietf-quic-transport.md]: Are cancelled or reset different?

File: draft-ietf-quic-transport.md (hunk 1 above)
Comment (ekr) [reply draft-ietf-quic-transport.md]: I would think should

File: draft-ietf-quic-transport.md (hunk 2 above)
Comment (ekr) [reply draft-ietf-quic-transport.md]: It's probably useful to say why you would do this.

File: draft-ietf-quic-transport.md [hunk 3]
```diff
@@ -2648,6 +2650,10 @@ closes a stream SHOULD be discarded.  An endpoint MAY choose to limit the period
 over which it ignores frames and treat frames that arrive after this time as
 being in error.
 
+An endpoint may receive a RST_STREAM after a FIN if the peer resets the stream
+after sending a FIN on it. In this case, the endpoint MAY discard any data that
```
Comment (ekr) [inline draft-ietf-quic-transport.md]: I would put the "if first" so it reads "If the peer resets the stream...."

File: draft-ietf-quic-transport.md (hunk 2 above)
Comment (ekr) [reply draft-ietf-quic-transport.md]: So I can also send multiple RST_STREAMs?

Comment (janaiyengar) [reply draft-ietf-quic-transport.md]: It can, but they'd be indistinguishable from a retransmission

**Comment by janaiyengar** [comment]
ekr: github is acting weird, so responding to one of your comments here. You suggested SHOULD for "the endpoint MAY discard any data that it already received on that stream." I'm sticking with MAY, since that's consistent with text around what to do with a received RST_STREAM. Perhaps we should change this in both places.

**Comment by MikeBishop** [comment]
This PR covers half-closed, but you can also send/receive RST_STREAM in closed to avert retransmissions.  If we're being explicit, we should cover that case as well.

**Comment by janaiyengar** [comment]
Thanks, Mike -- good point! I've added some text.

**Comment by MikeBishop** [comment]
👍

**Comment by janaiyengar** [comment]
Since this is simply stating explicitly what is already implicitly said, I am committing this. I am not closing the issue with this PR.

**Comment by janaiyengar** [comment]
Hm. I spoke too soon - I don't have any approvals on this PR yet, and I don't want to submit this with approval, so I'll wait for @martinthomson to review this.

File: draft-ietf-quic-transport.md [hunk 4]
```diff
@@ -2618,6 +2618,11 @@ on new STREAM frames.  Retransmission of data that has already been sent on
 STREAM frames is permitted.  An endpoint MAY also send MAX_STREAM_DATA and
 STOP_SENDING in this state.
 
+An application with a stream in this state may no longer care about data sent on
+the stream and may choose to reset the stream. To permit this behavior, a
```
Comment (martinthomson) [inline draft-ietf-quic-transport.md]: s/To permit this behavior, a/A/

Or just replace the first two sentences with "An application can decide to abandon a stream in this state.  An endpoint can send RST_STREAM for a stream that was closed with the FIN flag."  Keep the final offset piece.

**Comment by martinthomson** [comment]
Thanks Jana.

**Pull request merged**